### ISSUE TITLE
Add asynchronous counterpart operator of Finally: AsyncFinally

### ISSUE THREAD

**Issue by Weldryn** [dotnet/reactive#1215]
#### Feature request: AsyncFinally

> Which next library version (i.e., patch, minor or major)?

Next minor/major version?

> What are the platform(s), environment(s) and related component version(s)?

I would assume this to be as widely supported as async code in a finally can be.

> How commonly is this feature needed (one project, several projects, company-wide, global)?

I would assume this feature to be used globally anywhere async code is used. Some people already seems to use an operator functionnally equivalent [(SO link)](https://stackoverflow.com/a/45217578).

> Please describe the feature.

Add an operator equivalent to `Finally` in order to run an asynchronous delegate like `Func<Task>` as a final . Support for cancellation could be discussed.

Proposed signature would be, as per the SO answer:
`public static IObservable<TSource> AsyncFinally<TSource>(this IObservable<TSource> source, Func<Task> asyncAction);`

**Comment by quinmars** [comment]
The SO implementation is unfortunately not functional equivalent. The `Finally` operator calls the action while the observer is being disposed. A disposed operator will not propagate any new values or exceptions. Hence there is no point in awaiting the task. Well, it's an interesting question, however, where an exception thrown during `Finally` will end up. Probably it's thrown on whoever calls the `Dipose` method. Not a place where you expect this exception.

To sum it up, if your `Finally` action will not throw it is probably safe to not await the task (fire and forget). If you expect potential exceptions, I'm afraid, you have to re-think the usage of the `Finally` operator, even in the synchronous case.

**Comment by bartdesmet** [comment]
@quinmars is right; this doesn't compose well given the synchronous nature of the interfaces involved. It's something that Async Rx can (and does) provide though. Over there, the async nature can spread throughout the stack and bubble up all the way to the interfaces.
